File: swift/Sources/Golf/Lz4Block.swift
// Lz4Block.swift — self-contained LZ4 raw-block compressor and decompressor.
//
// Golf stores LZ4-compressed blocks in the raw LZ4 *block* format (not framed),
// prefixed with the uncompressed size as a little-endian uint32 — the same
// "prepend size" convention used by the reference implementations' `lz4_flex`
// bindings, Rust included. The prefix lives at the golf layer (see Core.swift);
// everything here operates on bare LZ4 block payloads.

/// Minimal-but-conformant implementation of the LZ4 block format.
///
/// The decoder parses arbitrary conformant blocks. The encoder uses a greedy
/// single-slot hash-chain matcher and obeys the format's safety rules the same
/// way the reference encoder does: minimum match length 4, offsets 1...65535,
/// no match starting within the last `mflimit` (12) bytes of the input, and
/// match extension capped so the final `lastLiterals` (5) bytes are always
/// emitted as a literal run — strict decoders such as liblz4 reject blocks
/// that end inside a match. Its ratio trails the reference `liblz4`, but the
/// output is bit-compatible with anything that speaks the block format.
enum Lz4Block {
    private static let minMatch = 4
    private static let mflimit = 12       // matches may not start in the last 12 bytes
    private static let lastLiterals = 5
    private static let hashLog = 14
    private static let hashSize = 1 << hashLog

    /// Worst-case compressed size for `size` input bytes.
    static func compressBound(_ size: Int) -> Int {
        size + size / 255 + 16
    }

    // MARK: Compress

    /// Compresses `src` into raw LZ4 block bytes.
    ///
    /// Always makes progress: for incompressible or tiny inputs the result is a
    /// single all-literals sequence, which is itself a valid block.
    static func compress(_ src: [UInt8]) -> [UInt8] {
        let n = src.count
        var out: [UInt8] = []
        out.reserveCapacity(compressBound(n))

        if n == 0 {
            out.append(0x00)
            return out
        }

        var table = [Int32](repeating: -1, count: hashSize)
        var anchor = 0   // start of pending literals
        var i = 0        // forward scan position

        while n - i >= mflimit + minMatch {
            let h = hash4(src, i)
            let candidate = table[h]
            // Record the current position immediately (before any `i`
            // mutation below) — a deferred write would observe the mutated i.
            table[h] = Int32(i)

            var tookMatch = false
            if candidate >= 0 {
                let cand = Int(candidate)
                let distance = i - cand
                if distance >= 1 && distance <= 65_535,
                   src[cand] == src[i],
                   src[cand + 1] == src[i + 1],
                   src[cand + 2] == src[i + 2],
                   src[cand + 3] == src[i + 3] {
                    // Match confirmed; extend like the reference encoder, whose
                    // match limit excludes the final `lastLiterals` bytes so the
                    // block always ends with a >=5-byte literal run (strict
                    // decoders such as liblz4 reject anything shorter).
                    var matchLen = minMatch
                    while i + matchLen < n - lastLiterals,
                          src[cand + matchLen] == src[i + matchLen] {
                        matchLen += 1
                    }

                    emitSequence(&out, src, anchor..<i, distance: distance, matchLen: matchLen)
                    anchor = i + matchLen
                    i = anchor
                    tookMatch = true
                }
            }

            if !tookMatch {
                i += 1
            }
        }

        // Trailing bytes become the mandatory final literal run.
        emitLastLiterals(&out, src, anchor..<n)
        return out
    }

    private static func hash4(_ src: [UInt8], _ i: Int) -> Int {
        let v = UInt32(src[i])
            | (UInt32(src[i + 1]) << 8)
            | (UInt32(src[i + 2]) << 16)
            | (UInt32(src[i + 3]) << 24)
        // Knuth's multiplicative hash; keep the top `hashLog` bits.
        let hashed = (v &* 2_654_435_761) >> (32 - hashLog)
        return Int(hashed)
    }

    /// Writes one [literals][match] sequence given its leading literal range.
    private static func emitSequence(
        _ out: inout [UInt8], _ src: [UInt8], _ literals: Range<Int>,
        distance: Int, matchLen: Int
    ) {
        precondition(matchLen >= minMatch)
        let litLen = literals.count
        let lenCode = matchLen - minMatch

        var token: UInt8 = litLen < 15 ? UInt8(litLen) << 4 : 0xF0
        if lenCode < 15 {
            token |= UInt8(lenCode)
        } else {
            token |= 0x0F
        }

        // Wire order: token | literal-length ext | literals | offset | match-length ext.
        out.append(token)
        if litLen >= 15 { writeExtendedLength(&out, litLen - 15) }
        out.append(contentsOf: src[literals])
        appendLE16(&out, UInt16(distance))
        if lenCode >= 15 { writeExtendedLength(&out, lenCode - 15) }
    }

    private static func writeExtendedLength(_ out: inout [UInt8], _ remaining: Int) {
        var rem = remaining
        while rem >= 255 {
            out.append(255)
            rem -= 255
        }
        out.append(UInt8(rem))
    }

    private static func appendLE16(_ out: inout [UInt8], _ v: UInt16) {
        out.append(UInt8(v & 0xFF))
        out.append(UInt8(v >> 8))
    }

    /// Writes the mandatory trailing literal-only sequence.
    private static func emitLastLiterals(_ out: inout [UInt8], _ src: [UInt8], _ range: Range<Int>) {
        let runLen = range.count
        if runLen >= 15 {
            out.append(0xF0)
            writeExtendedLength(&out, runLen - 15)
        } else {
            out.append(UInt8(runLen) << 4)
        }
        out.append(contentsOf: src[range])
    }

    // MARK: Decompress

    /// Decompresses a raw LZ4 block. `outputSize` bounds the destination; use
    /// the size recorded next to the block in a golf file.
    static func decompress(_ src: [UInt8], outputSize: Int) throws -> [UInt8] {
        var dst = [UInt8](repeating: 0, count: outputSize)
        var ip = 0      // input cursor
        var op = 0      // output cursor

        while ip < src.count {
            let token = src[ip]
            ip += 1

            // ── Literals ──
            var litLen = Int(token >> 4)
            if litLen == 15 {
                var extra: Int
                repeat {
                    guard ip < src.count else { throw GolfError.compressionError("truncated LZ4 literal length") }
                    extra = Int(src[ip]); ip += 1
                    litLen += extra
                    if litLen > outputSize { throw GolfError.compressionError("LZ4 literal length overflow") }
                } while extra == 255
            }
            guard ip + litLen <= src.count else { throw GolfError.compressionError("truncated LZ4 literals") }
            guard op + litLen <= outputSize else { throw GolfError.compressionError("LZ4 output overflow") }
            if litLen > 0 {
                dst.replaceSubrange(op..<op + litLen, with: src[ip..<ip + litLen])
                op += litLen
                ip += litLen
            }
            guard ip < src.count else { break } // final sequence ends with literals

            // ── Match ──
            guard ip + 2 <= src.count else { throw GolfError.compressionError("truncated LZ4 match offset") }
            let offset = Int(src[ip]) | (Int(src[ip + 1]) << 8)
            ip += 2
            guard offset >= 1 && offset <= op else { throw GolfError.compressionError("invalid LZ4 offset \(offset)") }

            var matchLen = Int(token & 0x0F) + minMatch
            if (token & 0x0F) == 15 {
                var extra: Int
                repeat {
                    guard ip < src.count else { throw GolfError.compressionError("truncated LZ4 match length") }
                    extra = Int(src[ip]); ip += 1
                    matchLen += extra
                    if matchLen > outputSize { throw GolfError.compressionError("LZ4 match length overflow") }
                } while extra == 255
            }
            guard op + matchLen <= outputSize else { throw GolfError.compressionError("LZ4 output overflow") }

            // Offsets smaller than the match length mean overlap: copy forward
            // one byte at a time so earlier output feeds later bytes.
            var sourcePos = op - offset
            var end = op
            while end < op + matchLen {
                dst[end] = dst[sourcePos]
                sourcePos += 1
                end += 1
            }
            op += matchLen
        }

        guard op == outputSize else {
            throw GolfError.compressionError("LZ4 produced \(op) bytes, expected \(outputSize)")
        }
        return dst
    }
}

File: kotlin/src/main/kotlin/golf/Lz4Block.kt
// Lz4Block.kt — self-contained LZ4 raw-block compressor and decompressor.
//
// Golf stores LZ4-compressed blocks in the raw LZ4 *block* format (not
// framed). The 4-byte little-endian uncompressed-size prefix lives at the golf
// layer (see Core.kt / Writer.kt); everything here operates on bare LZ4 block
// payloads.

package golf

/**
 * Minimal-but-conformant implementation of the LZ4 block format, ported from
 * the Swift implementation in this repository so both produce identical
 * sequences for identical input.
 *
 * The decoder parses arbitrary conformant blocks. The encoder uses a greedy
 * single-slot hash matcher and obeys the format's safety rules the same way
 * the reference encoder does: minimum match length 4, offsets 1..65535, no
 * match starting within the last `mflimit` (12) bytes of input, and match
 * extension capped so the final `lastLiterals` (5) bytes are always emitted
 * as a literal run — strict decoders such as liblz4 reject blocks that end
 * inside a match. Its ratio trails reference `liblz4`, but output is
 * compatible with anything that speaks the block format.
 */
object Lz4Block {
    private const val MIN_MATCH = 4
    private const val MFLIMIT = 12      // matches may not start in the last 12 bytes
    private const val LAST_LITERALS = 5 // final literal run guaranteed >= this
    private const val HASH_LOG = 14

    /** Worst-case compressed size for [size] input bytes. */
    fun compressBound(size: Int): Int = size + size / 255 + 16

    /** Compresses [src] into raw LZ4 block bytes. Always succeeds; tiny or
     *  incompressible inputs come back as a single all-literals sequence. */
    fun compress(src: ByteArray): ByteArray {
        val n = src.size
        if (n == 0) return byteArrayOf(0)

        val out = ByteArray(compressBound(n))
        var op = 0

        val table = IntArray(1 shl HASH_LOG) { -1 }
        var anchor = 0   // start of pending literals
        var i = 0        // forward scan position

        while (n - i >= MFLIMIT + MIN_MATCH) {
            val h = hash4(src, i)
            val candidate = table[h]
            // Record the current position immediately (before any mutation of
            // i below) so later candidates compare against the right offset.
            table[h] = i

            if (candidate >= 0 && candidate < i) {
                val cand = candidate
                val distance = i - cand
                if (distance in 1..65535 &&
                    src[cand] == src[i] &&
                    src[cand + 1] == src[i + 1] &&
                    src[cand + 2] == src[i + 2] &&
                    src[cand + 3] == src[i + 3]
                ) {
                    // Match confirmed; extend like the reference encoder,
                    // whose match limit excludes the final LAST_LITERALS
                    // bytes so the block always ends with a >=5-byte literal
                    // run (strict decoders reject anything shorter).
                    var matchLen = MIN_MATCH
                    while (i + matchLen < n - LAST_LITERALS &&
                           src[cand + matchLen] == src[i + matchLen]) {
                        matchLen++
                    }

                    op = emitSequence(out, op, src, anchor, i, distance, matchLen)
                    anchor = i + matchLen
                    i = anchor
                    continue
                }
            }
            i++
        }

        // Trailing bytes become the mandatory final literal run.
        op = emitLastLiterals(out, op, src, anchor, n)
        return out.copyOf(op)
    }

    private fun hash4(src: ByteArray, i: Int): Int {
        val v = (src[i].toUInt().and(0xFFu)) or
            ((src[i + 1].toUInt().and(0xFFu)) shl 8) or
            ((src[i + 2].toUInt().and(0xFFu)) shl 16) or
            ((src[i + 3].toUInt().and(0xFFu)) shl 24)
        // Knuth's multiplicative hash; keep the top HASH_LOG bits.
        return ((v * 2654435761u) shr (32 - HASH_LOG)).toInt()
    }

    /** Wire order: token | literal-length ext | literals | offset | match ext. */
    private fun emitSequence(
        out: ByteArray, opIn: Int, src: ByteArray,
        litStart: Int, litEnd: Int, distance: Int, matchLen: Int,
    ): Int {
        require(matchLen >= MIN_MATCH)
        var op = opIn
        val litLen = litEnd - litStart
        val lenCode = matchLen - MIN_MATCH

        var token = (if (litLen < 15) (litLen shl 4) else 0xF0)
        if (lenCode < 15) token = token or lenCode else token = token or 0x0F

        out[op++] = token.toByte()
        if (litLen >= 15) {
            var rem = litLen - 15
            while (rem >= 255) { out[op++] = 255.toByte(); rem -= 255 }
            out[op++] = rem.toByte()
        }
        System.arraycopy(src, litStart, out, op, litLen); op += litLen
        out[op++] = (distance and 0xFF).toByte()
        out[op++] = ((distance shr 8) and 0xFF).toByte()
        if (lenCode >= 15) {
            var rem = lenCode - 15
            while (rem >= 255) { out[op++] = 255.toByte(); rem -= 255 }
            out[op++] = rem.toByte()
        }
        return op
    }

    /** Writes the mandatory trailing literal-only sequence. */
    private fun emitLastLiterals(out: ByteArray, opIn: Int, src: ByteArray, start: Int, end: Int): Int {
        var op = opIn
        val runLen = end - start
        if (runLen >= 15) {
            out[op++] = 0xF0.toByte()
            var rem = runLen - 15
            while (rem >= 255) { out[op++] = 255.toByte(); rem -= 255 }
            out[op++] = rem.toByte()
        } else {
            out[op++] = (runLen shl 4).toByte()
        }
        System.arraycopy(src, start, out, op, runLen)
        return op + runLen
    }

    /**
     * Decompresses a raw LZ4 block. [outputSize] bounds the destination; use
     * the uncompressed size recorded next to the block in a golf file.
     */
    fun decompress(src: ByteArray, offset: Int, length: Int, outputSize: Int): ByteArray {
        val dst = ByteArray(outputSize)
        var ip = offset          // input cursor
        val ipEnd = offset + length
        var op = 0               // output cursor

        while (ip < ipEnd) {
            val token = src[ip++].toInt() and 0xFF

            // ── Literals ──
            var litLen = token ushr 4
            if (litLen == 15) {
                do {
                    checkInput(ip < ipEnd, "truncated LZ4 literal length")
                    val extra = src[ip++].toInt() and 0xFF
                    litLen += extra
                    check(litLen <= outputSize) { "LZ4 literal length overflow" }
                } while (extra == 255)
            }
            checkInput(ip + litLen <= ipEnd, "truncated LZ4 literals")
            check(op + litLen <= outputSize) { "LZ4 output overflow" }
            if (litLen > 0) {
                System.arraycopy(src, ip, dst, op, litLen)
                op += litLen
                ip += litLen
            }
            if (ip >= ipEnd) break           // final sequence ends with literals

            // ── Match ──
            checkInput(ip + 2 <= ipEnd, "truncated LZ4 match offset")
            val off = (src[ip].toInt() and 0xFF) or ((src[ip + 1].toInt() and 0xFF) shl 8)
            ip += 2
            if (off < 1 || off > op) throw GolfException("invalid LZ4 offset $off")

            var matchLen = (token and 0x0F) + MIN_MATCH
            if ((token and 0x0F) == 15) {
                do {
                    checkInput(ip < ipEnd, "truncated LZ4 match length")
                    val extra = src[ip++].toInt() and 0xFF
                    matchLen += extra
                    check(matchLen <= outputSize) { "LZ4 match length overflow" }
                } while (extra == 255)
            }
            check(op + matchLen <= outputSize) { "LZ4 output overflow" }

            // Offsets smaller than the match length mean overlap: copy forward
            // one byte at a time so earlier output feeds later bytes.
            var sp = op - off
            repeat(matchLen) { dst[it + op] = dst[sp++] }
            op += matchLen
        }

        if (op != outputSize) {
            throw GolfException("LZ4 produced $op bytes, expected $outputSize")
        }
        return dst
    }

    private fun checkInput(condition: Boolean, message: String) {
        if (!condition) throw GolfException(message)
    }
}
